src/System.Memory/tests/MemoryPool/MemoryPool.cs

```diff
@@ -0,0 +1,239 @@
+            for (int i = 0; i < 100; i++)
+            {
+                testInputs.Add((Math.Abs(r.Next() % 1000)) + 1);
+            }
+            testInputs.Add(pool.MaxBufferSize);
+
+            foreach (int minBufferSize in testInputs)
+            {
+                OwnedMemory<int> newBlock = pool.Rent(minBufferSize);
+                Assert.True(newBlock.Length >= minBufferSize);
+
+                foreach (OwnedMemory<int> prior in priorBlocks)
+                {
+                    using (MemoryHandle priorMemoryHandle = prior.Pin())
+                    {
+                        using (MemoryHandle newMemoryHandle = newBlock.Pin())
+                        {
+                            unsafe
+                            {
+                                Assert.NotEqual((IntPtr)priorMemoryHandle.Pointer, (IntPtr)newMemoryHandle.Pointer);
+                            }
+                        }
+                    }
+                }
+                priorBlocks.Add(newBlock);
+            }
+
+            foreach (OwnedMemory<int> prior in priorBlocks)
+            {
+                prior.Dispose();
+            }
+        }
+
+        [Fact]
+        public static void RentWithDefaultSize()
+        {
+            using (OwnedMemory<int> block = MemoryPool<int>.Shared.Rent(minBufferSize: -1))
+            {
+                Assert.True(block.Length >= 1);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(BadSizes))]
+        public static void RentBadSizes(int badSize)
+        {
+            MemoryPool<int> pool = MemoryPool<int>.Shared;
+            Assert.Throws<ArgumentOutOfRangeException>(() => pool.Rent(minBufferSize: badSize));
+        }
+
+        public static IEnumerable<object[]> BadSizes
+        {
+            get
+            {
+                yield return new object[] { 0 };
+                yield return new object[] { -2 };
+                yield return new object[] { int.MinValue };
+                yield return new object[] { int.MaxValue };
+                yield return new object[] { MemoryPool<int>.Shared.MaxBufferSize + 1 };
+            }
+        }
+
+        [Fact]
+        public static void MemoryPoolTryGetArray()
+        {
+            using (OwnedMemory<int> block = MemoryPool<int>.Shared.Rent(42))
+            {
+                Memory<int> memory = block.Memory;
+                bool success = memory.TryGetArray(out ArraySegment<int> arraySegment);
+                Assert.True(success);
+                Assert.Equal(block.Length, arraySegment.Count);
+                unsafe
+                {
+                    void* pSpan = Unsafe.AsPointer(ref MemoryMarshal.GetReference(block.Span));
+                    fixed (int* pArray = arraySegment.Array)
+                    {
+                        Assert.Equal((IntPtr)pSpan, (IntPtr)pArray);
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public static void RefCounting()
```

Comment:
Could you add some stress tests too to ensure we don't let people access the array (the Span) after Dispose on a separate thread. Also, to ensure that Dispose throws is there are outstanding Retain (issues on a different thread).

Reply:
>Dispose on a separate thread

That would be contingent on whether we want to guarantee thread-safety. I'm inclined against that.

On a separate note, is CI really the place to be doing stress tests? CI should be about prevent clearly bad stuff from going in, not randomly holding up PR's because of intermittent stress failures.

The Dispose throwing on an outstanding ref-count is already tested in the OwnedMemory tests (and the OwnedMemory class is the one that does that check so checking for individual variants doesn't add any code coverage.) It's an invariant of OwnedMemory objects in general and not specific to MemoryPool.